### PR TITLE
Fix ToolStripItemCollection.AddRange ArgumentOutOfRangeException

### LINKED ISSUE #4454: ToolStripItemCollection.AddRange(ToolStripItemCollection toolStripItems) causes ArgumentOutOfRangeException as it deletes from source collection
This method cannot be used as it will always cause an ArgumentOutOfRangeException since items are deleted from the source collection while they are copied into the new collection.

Take this code in Form_Load()
    ContextMenuStrip1.Items.AddRange(EditToolStripMenuItem.DropDownItems)

Before it is executed, ContextMenuStrip1 has 0 items and EditToolStripMenuItem has 16. While it is called and the exception is triggered, both have 8 but alternating. I.e. ContextMenuStrip1 now has indexes 0,2,4,6,8,10,12,14 from the original EditToolStripMenuItem. EditToolStripMenuItem is left with 1,3,5,7,9,11,13,15.

The exception occurs because it is expected there will be 16 items to move, but only 8 exist in the array now.

Expected behavior: Each reference is copied into the target menu. Not removed from source.

Watches before: [https://pasteboard.co/JJhkD66.png](url)
After (at exception): [https://pasteboard.co/JJhkV9w.png](url)
(Apologies for random image site. Most are blocked by our corporate firewall)

This was tested on VB .NET 4.6.

I already have a workaround.

### PR DESCRIPTION
**Pull request by willibrandon**
Fixes #4454 

## Proposed changes

- Fix the ToolStripItemCollection.AddRange ArgumentOutOfRangeException.
- Add a test to validate the fix.

As @weltkante [mentioned here](https://github.com/dotnet/winforms/issues/4454#issuecomment-759381740), the only non-breaking option is to implement `AddRange` in a way that moves all of the items without throwing an exception. There are several ways to solve this problem and I would appreciate any feedback.

The approach that I've taken here is to decrement the index to always `Add` from index 0 which preserves the original order, does not require reversing the collection, and avoids the ArgumentOutOfRangeException.

## Customer Impact

- Fixes the ArgumentOutOfRangeException and adds the elements to the end of the ToolStripItemCollection inner ArrayList.

## Regression? 

- No.

## Risk

- Could break people who expect to catch the ArgumentOutOfRangeException but this is unlikely and not something that should be considered breaking.

## Test methodology 

- Added a new unit test.
- Ran unit tests.

## Test environment(s) 

```
.NET SDK (reflecting any global.json):
 Version:   5.0.100
 Commit:    5044b93829

Runtime Environment:
 OS Name:     Windows
 OS Version:  10.0.19042
 OS Platform: Windows
 RID:         win10-x64
 Base Path:   C:\Program Files\dotnet\sdk\5.0.100\

Host (useful for support):
  Version: 5.0.2
  Commit:  cb5f173b96

.NET SDKs installed:
  5.0.100-rc.1.20452.10 [C:\Program Files\dotnet\sdk]
  5.0.100 [C:\Program Files\dotnet\sdk]
  5.0.102 [C:\Program Files\dotnet\sdk]
  5.0.200-preview.20601.7 [C:\Program Files\dotnet\sdk]

.NET runtimes installed:
  Microsoft.AspNetCore.All 2.1.23 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.All]
  Microsoft.AspNetCore.All 2.1.24 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.All]
  Microsoft.AspNetCore.App 2.1.23 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 2.1.24 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 3.1.9 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 3.1.11 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 5.0.0-rc.1.20451.17 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 5.0.0 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 5.0.2 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.NETCore.App 2.1.23 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 2.1.24 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 3.1.9 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 3.1.11 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 5.0.0-rc.1.20451.14 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 5.0.0 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 5.0.2 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.WindowsDesktop.App 3.1.9 [C:\Program Files\dotnet\shared\Microsoft.WindowsDesktop.App]
  Microsoft.WindowsDesktop.App 3.1.11 [C:\Program Files\dotnet\shared\Microsoft.WindowsDesktop.App]
  Microsoft.WindowsDesktop.App 5.0.0-rc.1.20452.2 [C:\Program Files\dotnet\shared\Microsoft.WindowsDesktop.App]
  Microsoft.WindowsDesktop.App 5.0.0-rc.2.20475.6 [C:\Program Files\dotnet\shared\Microsoft.WindowsDesktop.App]
  Microsoft.WindowsDesktop.App 5.0.0 [C:\Program Files\dotnet\shared\Microsoft.WindowsDesktop.App]
  Microsoft.WindowsDesktop.App 5.0.2 [C:\Program Files\dotnet\shared\Microsoft.WindowsDesktop.App]
```

###### Microsoft Reviewers: [Open in CodeFlow](http://wpcp.azurewebsites.net/CodeFlowProtocolProxy2.php?pullrequest=https://github.com/dotnet/winforms/pull/4481)

### CODE REVIEW

**Comment by RussKie** [comment]
A gentle nudge to see whether this is ready for review. Thanks

**Comment by willibrandon** [comment]
Yes I'm curious what the teams thoughts are on this. However, I understand that since this is not a regression issue, that it is not a high priority.

**Comment by RussKie** [comment]
You marked it as a draft, so I wasn't sure if it was ready for a review. I'm OOF, will review next week when I'm back.
Thanks
________________________________

**Review by RussKie — COMMENTED** [review]
👍

File: src/System.Windows.Forms/src/System/Windows/Forms/ToolStripItemCollection.cs [hunk 1]
```diff
@@ -149,10 +149,9 @@ public void AddRange(ToolStripItemCollection toolStripItems)
             // to resize itself.
             using (new LayoutTransaction(_owner, _owner, PropertyNames.Items))
             {
-                int currentCount = toolStripItems.Count;
-                for (int i = 0; i < currentCount; i++)
+                for (int i = 0; i < toolStripItems.Count; i++)
                 {
-                    Add(toolStripItems[i]);
+                    Add(toolStripItems[i--]);
```
Comment (dreddy-work) [inline src/System.Windows.Forms/src/System/Windows/Forms/ToolStripItemCollection.cs:154]: Can you add comment here for future reference on why we are decrementing this?

**Review by RussKie — DISMISSED** [review]

File: src/System.Windows.Forms/src/System/Windows/Forms/ToolStripItemCollection.cs (hunk 1 above)
Comment (willibrandon) [reply src/System.Windows.Forms/src/System/Windows/Forms/ToolStripItemCollection.cs:157]: Thank you. Done 👍

**Review by dreddy-work — APPROVED** [review]
LGTM.

**Review by RussKie — APPROVED** [review]

**Comment by RussKie** [comment]
Thank you!

**Pull request merged**